exporter/elasticsearchexporter/model.go

```diff
@@ -155,6 +161,126 @@ func (m *encodeModel) encodeLogECSMode(resource pcommon.Resource, record plog.Lo
 	return document
 }
 
+func (m *encodeModel) encodeMetrics(resource pcommon.Resource, metrics pmetric.MetricSlice, _ pcommon.InstrumentationScope) ([][]byte, error) {
+	var baseDoc objmodel.Document
+
+	baseDoc.AddAttributes("", resource.Attributes())
+
+	// Put all metrics that have the same attributes and timestamp in one document.
+	docs := map[uint32]*objmodel.Document{}
+	for i := 0; i < metrics.Len(); i++ {
+		metric := metrics.At(i)
+
+		var dps pmetric.NumberDataPointSlice
+
+		// Only Gauge and Sum metric types are supported at the moment.
+		switch metric.Type() {
+		case pmetric.MetricTypeGauge:
+			dps = metric.Gauge().DataPoints()
+		case pmetric.MetricTypeSum:
+			dps = metric.Sum().DataPoints()
+		}
+
+		for j := 0; j < dps.Len(); j++ {
+			dp := dps.At(j)
+
+			hash := metricHash(dp.Timestamp(), dp.Attributes())
+			doc, docExists := docs[hash]
+			if !docExists {
+				doc = baseDoc.Clone()
+				doc.AddTimestamp("@timestamp", dp.Timestamp())
+				doc.AddAttributes("", dp.Attributes())
+
+				docs[hash] = doc
+			}
+
+			switch dp.ValueType() {
+			case pmetric.NumberDataPointValueTypeDouble:
+				doc.AddAttribute(metric.Name(), pcommon.NewValueDouble(dp.DoubleValue()))
+			case pmetric.NumberDataPointValueTypeInt:
+				doc.AddAttribute(metric.Name(), pcommon.NewValueInt(dp.IntValue()))
+			}
+		}
+	}
+
+	res := make([][]byte, 0, len(docs))
+
+	for _, doc := range docs {
+		if m.dedup {
+			doc.Dedup()
+		} else if m.dedot {
+			doc.Sort()
+		}
+
+		var buf bytes.Buffer
+		err := doc.Serialize(&buf, m.dedot)
+
+		if err != nil {
+			return nil, err
+		} else {
+			res = append(res, buf.Bytes())
+		}
+	}
+
+	return res, nil
+}
+
+func metricHash(timestamp pcommon.Timestamp, attributes pcommon.Map) uint32 {
+	hasher := fnv.New32a()
+
+	timestampBuf := make([]byte, 8)
+	binary.LittleEndian.PutUint64(timestampBuf, uint64(timestamp))
+	hasher.Write(timestampBuf)
+
+	mapHash(hasher, attributes)
+
+	return hasher.Sum32()
+}
+
+func mapHash(hasher hash.Hash, m pcommon.Map) {
+	m.Range(func(k string, v pcommon.Value) bool {
+		hasher.Write([]byte(k))
+		valueHash(hasher, v)
+
+		return true
+	})
+}
+
+func valueHash(h hash.Hash, v pcommon.Value) {
+	switch v.Type() {
+	case pcommon.ValueTypeEmpty:
+		h.Write([]byte{0})
+	case pcommon.ValueTypeStr:
+		h.Write([]byte(v.Str()))
+	case pcommon.ValueTypeBool:
+		if v.Bool() {
+			h.Write([]byte{1})
+		} else {
+			h.Write([]byte{0})
+		}
+	case pcommon.ValueTypeDouble:
+		buf := make([]byte, 8)
+		binary.LittleEndian.PutUint64(buf, math.Float64bits(v.Double()))
+		h.Write(buf)
+	case pcommon.ValueTypeInt:
+		buf := make([]byte, 8)
+		binary.LittleEndian.PutUint64(buf, uint64(v.Int()))
+		h.Write(buf)
+	case pcommon.ValueTypeBytes:
+		h.Write(v.Bytes().AsRaw())
+	case pcommon.ValueTypeMap:
+		mapHash(h, v.Map())
+	case pcommon.ValueTypeSlice:
+		sliceHash(h, v.Slice())
+	}
+}
+
+func sliceHash(h hash.Hash, s pcommon.Slice) {
+	for i := 0; i < s.Len(); i++ {
+		valueHash(h, s.At(i))
+	}
+}
```

Comment:
I just came across https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/pdatautil/hash.go, which does a lot of this already (but using xxhash). Perhaps we should extend and use that?

I think the main thing we would be missing from pdatautil is the ability to keep state while hashing multiple values -- timestamp & attributes in our case. We could alternatively key on (timestamp, hash(attributes)).

Lastly, we could alternatively build on https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/internal/exp/metrics/identity (specifically "OfStream"). But again, that doesn't include timestamp...

Reply:
Thanks for finding pdatautil, I looked briefly for something like this but didn't see it. Definitely worth looking into it.

Reply:
Due to a change in the implementation, the whole hashing part is now  removed.